add some notes about all the different zeros (#162)
JuliaDiff/ChainRulesCore.jl: docs/src/design/zeros.md

# Design Notes: The many kinds of Zeros and NotDefined's

> There are many zero and not defined line situtions one might want to talk about in the context of differentation.

Reviewer: What's a "line situation"?

Author: I don't know

Not all of them can be generated by autodiff software.
Not all of them are supported by ChainRules.

Here is list of some of the examples:

> Differentials are roughtly a vector-field they support scaling.
> So there is at least 1 scalar zero that much be supported.
> But one might want an extra one that can resolve at compile time based on type, and avoid `unthunk`ing `Thunks`, to save on computation.
> (Or one might want to bake that into the notion of `*::(x, t::AbstractThunk) = iszero(x) ? ...`).

Reviewer: ```suggestion
Differentials are roughly a vector-space -- they support scaling.
So there is at least 1 scalar zero that must be supported.
But one might want an extra zero that can resolved at compile time based on type to completely avoid certain computations e.g. `unthunk`ing `Thunks`.
(Or one might want to bake that into the notion of `*::(x, t::AbstractThunk) = iszero(x) ? ...`).
```

And this can be stacked with the limit differing cases mentioned earlier, so the primal function is not defined and the limit from each side does not agree but either encloses or does not enclose zero.

There is the zero that is `\dfrac{\partial f}{\partial a}` for ``f(a,b)=2b``.
This one is particular important I feel in source to source AD.
It represents a disconnection in the computational graph, there is no path from input ``a`` to the output ``f(a,b)``.
This one can also show up dynamically, but perhaps that should be considered a different case.
For example in `max(a,b)` or in ``ifelse(cond, a, b)``.
Have has a few talks with [James Bradbury](https://github.com/jekbradbury) about this, apparently it is important this this is a strong-zero, like julia's `false` where `false*NaN=false` not `NaN`.
This is the subject of TensorFlow's _double where trick_, (`where` is what `ifelse` is called in TensorFlow) as they do not have a strong-zero.
If a gradient being propagated backwards from a branch that was not taken is `NaN`, and thus the `ifelse` has this disconnected zero,  then when the chainrule is applied it is required that this zero remains zero (not `NaN`).

> I have not seen a good writeup on this, apprently one exists somewhere in the TensorFlow issue tracker.

Reviewer: ```suggestion
I have not seen a good writeup on this, apparently one exists somewhere in the TensorFlow issue tracker.
```

There is the zero/not-definedness for something where perturbing its value is an error.
So this is the gradient of `f'(5)` for ``f(x) = [1,2,3,4][x]``.
As small perturbation to this is an error, e.g. `f(5.1)` is not defined.
Related to that is where the notion of perterbing is not defined.
This is the case is for inputs that are `String`s or `Symbol`s.

> There is the cases of a structural Zero is a sparse data structure.

Reviewer: Not sure what you're saying here.

Author: ```suggestion
There is the cases of a structural Zero in a sparse data structure.
```

Like the off-diagonal on a `DiagonalMatrix`.

> Also the structual zero of a `SparseCSC` that varies at run time, particularly relevent in as it can be the result from the derivative of `getindex`.

Reviewer: ```suggestion
Also the structural zero of a `SparseCSC` that varies at run time, particularly relevant in that it can be the result from the derivative of `getindex`.
```
?

As well as the zero that could be within the differential representing a tuple
if it is ``f(x::Tuple{Float64,Float64,Float64,}) = x[1] + x[3]``
then derivative is `Composite{Tuple}(1, Zero(), 1)` and that is a structual zero.

Derivative with repect to empty things.

> They have no value so can not be perturbated.

Reviewer: ```suggestion
They have no value so can not be perturbed.
```

For example the gradient with respect to a empty array or tuple.
Also with respect to an struct that has no fields.

> The struct case is interesting as a struct without fields is a singlton

Reviewer: ```suggestion
The struct case is interesting as a struct without fields is a singleton
```

> (technically a `mutable struct` isn't but it migth as well be).

Reviewer: ```suggestion
(technically a `mutable struct` isn't but it might as well be).
```

It is the only the only element of its type.
A very common case of this is functions.
Every function in julia is a singlton struct, with call overloading.

> This is ChainRules's `Δself` that shows up in pullbacks and pushforward -- it is this kind of zero whenver

Reviewer: I think you stopped writing here.
